### PR TITLE
[HttpFoundation] overwrite default encodingOptions in child class

### PR DESCRIPTION
**Pull request by alexpts**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | no |
| License | MIT |
| Doc PR |  |

This will allow to change the default encoding options without calling `\json_decode` and `\json_encode`.

### CODE REVIEW

File: src/Symfony/Component/HttpFoundation/JsonResponse.php [hunk 1]
```diff
@@ -43,12 +43,17 @@ public function __construct($data = null, $status = 200, $headers = array())
...
+    
+    protected function setDefaultEncodingOptions()
+    {
         // Encode <, >, ', &, and " for RFC4627-compliant JSON, which may also be embedded into HTML.
         $this->encodingOptions = JSON_HEX_TAG | JSON_HEX_APOS | JSON_HEX_AMP | JSON_HEX_QUOT;
-
-        $this->setData($data);
     }
-
+
```
Comment (hhamon) [inline src/Symfony/Component/HttpFoundation/JsonResponse.php]: the indent must be removed.

Comment (alexpts) [inline src/Symfony/Component/HttpFoundation/JsonResponse.php]: @hhamon, thanks!

File: src/Symfony/Component/HttpFoundation/JsonResponse.php [hunk 2]
```diff
@@ -43,10 +43,14 @@ public function __construct($data = null, $status = 200, $headers = array())
             $data = new \ArrayObject();
         }
 
+        $this->setDefaultEncodingOptions();
+        $this->setData($data);
+    }
+
+    protected function setDefaultEncodingOptions()
```
Comment (fabpot) [inline src/Symfony/Component/HttpFoundation/JsonResponse.php]: It means that you need to create a sub-class to be able to override the encoding options. What about creating a real setter instead?

Comment (alexpts) [inline src/Symfony/Component/HttpFoundation/JsonResponse.php]: Class has setter method, but it is no optimal
```php
public function setEncodingOptions($encodingOptions)
    {
        $this->encodingOptions = (int) $encodingOptions;

        return $this->setData(json_decode($this->data));
    }
```

When I need use not default encodingOptions code will work not optimally:
json_encode (constuctor -> setData) -> json_decode (setEncodingOptions) -> json_encode (setEncodingOptions -> setData)

I think better way could be fourth optional parameter to constructor with current hardcoded value encodingOptions as default, but expressions is disallow in function parameter.

Therefore, I see overwrite in the child class with the current implementation.

Do you have any ideas how to do this without inheritance?

**Pull request closed without merging**